ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InputsAndEffects.scala

```diff
@@ -84,8 +85,10 @@ private[kvutils] object InputsAndEffects {
           case fetch: Node.NodeFetch[Value.ContractId] =>
             addContractInput(fetch.coid)
             fetch.key.foreach { keyWithMaintainers =>
-              inputs += globalKeyToStateKey(
+              val key = globalKeyToStateKey(
                 GlobalKey(fetch.templateId, forceNoContractIds(keyWithMaintainers.key)))
+              inputs += key
+              resolvedContractIds += resolvedContractKeyIdPair(key.getContractKey, fetch.coid)
```

Comment:
This looks fishy. The association of a contract key to a contract ID may change during a transaction, even several times. So overall we may get several `ContractKeyIdPair`s for the same key in the list.

Worse, you're iterating here in an unspecified order over the transaction nodes using `tx.foreach`, so it's not even clear that the pairs for the same key come in the right order. 

Moreover, `Create` nodes and negative `LookupByKey` nodes don't contribute to the key resolutions. So in a transaction that first does a negative `LookupByKey` for a key `k`, then `Create`s a contract `c` for the `k`, and then `FetchByKey` on `k`, we only get a resolution saying `k -> c`, even though at the beginning of the transaction, we have `k -> None` from the perspective of the submitter.

Reply:
I've come up to a similar conclusion yesterday but needed confirmation. Thanks @andreaslochbihler-da !
The fact that a simple `aKey -> aContractId` mapping is not sufficient means that we need a different way to encode how a participant resolved keys. I cannot see a good solution right away but will work on that.

Reply:
You don't need to ship all resolutions. It's enough if you ship the expected key resolutions at the beginning of the transaction and check that. DAML engine will take care of the rest. For that, you need to traverse the transaction in execution order, though, and record the first node's effect for each key.

Reply:
Changed the logic to address @andreaslochbihler-da comments:
- traversing the transaction tree in execution order
- shipping only the first node's effect for a key
- in case of negative key lookups an empty string is shipped as a contract id